plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx

```diff
@@ -186,32 +189,31 @@ export const Stepper = (stepperProps: StepperProps) => {
 
   const handleNext = useCallback(
     async ({ formData = {} }: { formData?: Record<string, JsonValue> }) => {
-      // The validation should never throw, as the validators are wrapped in a try/catch.
-      // This makes it fine to set and unset state without try/catch.
       setErrors(undefined);
       setIsValidating(true);
 
       const returnedValidation = await validation(formData);
 
-      setStepsState(current => ({
-        ...current,
-        ...formData,
-      }));
-
-      setIsValidating(false);
-
       if (hasErrors(returnedValidation)) {
         setErrors(returnedValidation);
-      } else {
-        setErrors(undefined);
-        setActiveStep(prevActiveStep => {
-          const stepNum = prevActiveStep + 1;
-          analytics.captureEvent('click', `Next Step (${stepNum})`);
-          return stepNum;
-        });
+        setIsValidating(false);
+        return;
       }
+
+      const cleanedFormData = { ...formData };
+      Object.keys(secrets).forEach(secretKey => {
+        delete cleanedFormData[secretKey];
```

Comment:
I think unfortunately, this only works for top level secret keys.

Imagine you have a schema like this:

```yaml
parameters:
  test:
    type: object
    required: [thing]
    properties:
      thing:
        ui:field: SecretField
```

Then this solution doesn't work right?

I think it's part of a bigger refactor of how we deal with the form state and secrets #28773